ImageData (rect) size and destination offset overflow causes IPC decoder to fail during 2D context rendering.

nishajain61 opened the pull request:
#### 80efddf753c773046959715f02cb7421cf8245ce
<pre>
ImageData (rect) size and destination offset overflow causes IPC decoder to fail during 2D context rendering.
<a href="https://bugs.webkit.org/show_bug.cgi?id=271080">https://bugs.webkit.org/show_bug.cgi?id=271080</a>
<a href="https://rdar.apple.com/121959787">rdar://121959787</a>

Reviewed by Kimmo Kinnunen.

ImageData (rect) size and destination offset overflow causes IPC (rect) object
decoder to fail in WebContentProcess, during 2D context rendering which causes timeout.
Using location and size params in the getPixelBuffer and putPixelBuffer APIs instead
of rect in order to avoid this issue.

* LayoutTests/fast/canvas/canvas-imagedata-dimension-no-overflow-expected.txt: Added.
* LayoutTests/fast/canvas/canvas-imagedata-dimension-no-overflow.html: Added.
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::getPixelBuffer):
(WebKit::RemoteImageBuffer::getPixelBufferWithNewMemory):
(WebKit::RemoteImageBuffer::putPixelBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::getPixelBufferForImageBuffer):
(WebKit::RemoteRenderingBackendProxy::putPixelBufferForImageBuffer):

Canonical link: <a href="https://commits.webkit.org/276563@main">https://commits.webkit.org/276563@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4bf2effe5ae7f9e50a7c54e0e60d70f8a549ab30

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44766 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23860 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47244 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47420 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40771 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27908 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21253 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36792 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45344 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20936 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38560 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17858 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18364 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39691 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2813 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41011 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39980 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49081 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19731 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16316 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43772 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21051 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/38791 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42519 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10013 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/21390 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20725 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on LayoutTests/fast/canvas/canvas-intrect-size-overflow.html:
```diff
@@ -0,0 +1,45 @@
+<head>
```
intrect isn't a HTML/DOM concept.
Maybe you could call it "canvas-imagedata-dimensions-no-overflow.html" or something

nishajain61 replied:
Fixed

kkinnunen-apple commented on LayoutTests/fast/canvas/canvas-intrect-size-overflow-expected.txt:
```diff
@@ -0,0 +1,2 @@
+CONSOLE MESSAGE: ReferenceError: Can't find variable: shouldBe
```
Remember to update the expectations

nishajain61 replied:
Fixed

kkinnunen-apple reviewed (commented):
>IntRect size overflow causes IPC (IntRect) object decoder to fail in WebContentProcess which causes timeout.

This line should summarize what your change is about.
It should probably mention the problem in a bit higher level (image data size and destination offset overflow), and it should mention the area which you are fixing (2d context).


> https://bugs.webkit.org/show_bug.cgi?id=271080.

No dots at the end of url

> [rdar://121959787](https://rdar.apple.com/121959787).

No dots at the end of url.

> IntRect size overflow causes IPC decoder to fail in WebContentProcess which causes timeout.
> Using location and size params in the getPixelBuffer and putPixelBuffer APIs instead of rect in order to avoid this issue.

Just a suggestion: it would be easier to read if you split the lines. You can spend time reading your own PR, so you see how it looks like.

kkinnunen-apple reviewed (commented):
> ImageData (rect) size and destination offset overflow causes
> IPC (rect) object decoder to fail in WebContentProcess during
> 2D context rendering which causes timeout.
https://bugs.webkit.org/show_bug.cgi?id=271080
[rdar://121959787](https://rdar.apple.com/121959787)

The commit title cannot be split into multiple lines. So you should just write one line, preferably not too long. So you might want to still cull the unneeded information out of the title and un-split that.

kkinnunen-apple commented on LayoutTests/fast/canvas/canvas-imagedata-dimension-no-overflow-expected.txt:
```diff
@@ -0,0 +1,24 @@
+CONSOLE MESSAGE: [object ImageData]
```
You don't need to connsole log anything like these.

nishajain61 replied:
Fixed

kkinnunen-apple commented on LayoutTests/fast/canvas/canvas-imagedata-dimension-no-overflow.html:
```diff
@@ -0,0 +1,45 @@
+console.log(result);
+var data = result.data;
+console.log('Drawing canvas with MaxSize and getting ImageData from image');
+// expect that d[0] == 0, d[1] == 255, d[2] ==  0, d[3] == 255;
```
This comment was meant for you to understand what you need to do.
Now that you have added the assert, you can remove it, it doesn't provide any clarification.

nishajain61 replied:
Fixed

kkinnunen-apple commented on LayoutTests/fast/canvas/canvas-imagedata-dimension-no-overflow.html:
```diff
@@ -0,0 +1,45 @@
+var result = ctx.getImageData(1, 2147483647, 1, 1);
+console.log(result);
+var data = result.data;
+// expect that d[0] == 0, d[1] == 0, d[2] ==  0, d[3] == 0;
```
This could be removed

nishajain61 replied:
Fixed

kkinnunen-apple commented on LayoutTests/fast/canvas/canvas-imagedata-dimension-no-overflow.html:
```diff
@@ -0,0 +1,45 @@
+ctx.putImageData(result, 1, 2147483647);
+result = ctx.getImageData(2147483646, 1, 1, 1);
+data = result.data;
+// expect that d[0] == 0, d[1] == 0, d[2] ==  0, d[3] == 0;
```
this could be removed.

nishajain61 replied:
Fixed

nishajain61 committed:
ImageData (rect) size and destination offset overflow causes IPC decoder to fail during 2D context rendering.
https://bugs.webkit.org/show_bug.cgi?id=271080
rdar://121959787

Reviewed by Kimmo Kinnunen.

ImageData (rect) size and destination offset overflow causes IPC (rect) object
decoder to fail in WebContentProcess, during 2D context rendering which causes timeout.
Using location and size params in the getPixelBuffer and putPixelBuffer APIs instead
of rect in order to avoid this issue.

* LayoutTests/fast/canvas/canvas-imagedata-dimension-no-overflow-expected.txt: Added.
* LayoutTests/fast/canvas/canvas-imagedata-dimension-no-overflow.html: Added.
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::getPixelBuffer):
(WebKit::RemoteImageBuffer::getPixelBufferWithNewMemory):
(WebKit::RemoteImageBuffer::putPixelBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::getPixelBufferForImageBuffer):
(WebKit::RemoteRenderingBackendProxy::putPixelBufferForImageBuffer):

Canonical link: https://commits.webkit.org/276563@main
